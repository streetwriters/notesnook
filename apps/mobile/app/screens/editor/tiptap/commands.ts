/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { Note } from "@notesnook/core";
import type {
  Attachment,
  ImageAttributes,
  LinkAttributes
} from "@notesnook/editor";
import { createRef, RefObject } from "react";
import { Platform } from "react-native";
import { EdgeInsets } from "react-native-safe-area-context";
import WebView from "react-native-webview";
import { db } from "../../../common/database";
import { sleep } from "../../../utils/time";
import { Settings } from "./types";
import { useTabStore } from "./use-tab-store";
import { getResponse, randId, textInput } from "./utils";

type Action = { job: string; id: string };

async function call(webview: RefObject<WebView | undefined>, action?: Action) {
  if (!webview.current || !action) return;
  setImmediate(() => webview.current?.injectJavaScript(action.job));
  const response = await getResponse(action.id);
  return response ? response.value : response;
}

const fn = (fn: string, name?: string) => {
  const id = randId("fn_");
  return {
    job: `(async () => {
      if (typeof __PLATFORM__ === "undefined") __PLATFORM__ = "${Platform.OS}";
      try {
        let response = true;
        ${fn}
        post("${id}",response);
      } catch(e) {
        const DEV_MODE = ${__DEV__};
        if (DEV_MODE && typeof logger !== "undefined") logger('error', "webview: ", e.message, e.stack, "${name}");
      }
      return true;
    })();true;`,
    id: id
  };
};

class Commands {
  ref = createRef<WebView | undefined>();
  previousSettings: Partial<Settings> | null;
  constructor(ref: RefObject<WebView>) {
    this.ref = ref;
    this.previousSettings = null;
  }

  async doAsync<T>(job: string, name?: string) {
    if (!this.ref.current) return false;
    return call(this.ref, fn(job, name)) as Promise<T>;
  }

  async sendCommand<T>(command: string, ...args: any[]) {
    return this.doAsync(
      `response = globalThis.commands.${command}(${args
        .map((arg) =>
          typeof arg === "string" ? `"${arg}"` : JSON.stringify(arg)
        )
        .join(",")})`,
      command
    );
  }

  focus = async (tabId: string) => {
    if (!this.ref.current) return;

    const locked = useTabStore.getState().getTab(tabId)?.session?.locked;
    if (Platform.OS === "android") {
      setTimeout(async () => {
        if (
          locked &&
          useTabStore.getState().biometryAvailable &&
          useTabStore.getState().biometryEnrolled
        )
          return;

        if (!this.ref) return;
        textInput.current?.focus();
        await this.sendCommand("focus", tabId, locked);
        this.ref?.current?.requestFocus();
      }, 1);
    } else {
      await sleep(400);
      await this.sendCommand("focus", tabId, locked);
    }
  };

  blur = async (tabId: string) => this.sendCommand("blur", tabId);

  clearContent = async (tabId: string) => {
    this.previousSettings = null;
    await this.sendCommand("clearContent", tabId);
  };

  setSessionId = async (id: string | null) =>
    await this.sendCommand("setSessionId", id);

  setStatus = async (
    date: string | undefined,
    saved: string,
    tabId: string
  ) => {
    this.sendCommand("setStatus", date, saved, tabId);
  };

  setPlaceholder = async (placeholder: string) => {};

  setLoading = async (loading?: boolean, tabId?: string) => {
    this.sendCommand(
      "setLoading",
      loading,
      tabId === undefined ? useTabStore.getState().currentTab : tabId
    );
  };

  setInsets = async (insets: EdgeInsets) => {
    this.sendCommand("setInsets", insets);
  };

  updateSettings = async (settings?: Partial<Settings>) => {
    if (!this.previousSettings) return;
    this.previousSettings = {
      ...this.previousSettings,
      ...settings
    };
    this.sendCommand("updateSettings", settings);
  };

  setSettings = async (settings?: Partial<Settings>) => {
    if (settings) {
      this.previousSettings = settings;
    } else {
      if (this.previousSettings) {
        settings = this.previousSettings;
      } else {
        return;
      }
    }
    this.sendCommand("setSettings", settings);
  };

  setTags = async (note: Note | null | undefined) => {
    if (!note) return;
    useTabStore.getState().forEachNoteTab(note.id, async (tab) => {
      const tabId = tab.id;
      const tags = await db.relations.to(note, "tag").resolve();
      await this.sendCommand("setTags", tabId, tags);
    });
  };

  clearTags = async (tabId: string) => {
    await this.sendCommand("clearTags", tabId);
  };

  insertAttachment = async (attachment: Attachment, tabId: string) => {
    await this.sendCommand("insertAttachment", attachment, tabId);
  };

  setAttachmentProgress = async (
    attachmentProgress: Partial<Attachment>,
    tabId: string
  ) => {
    await this.sendCommand("setAttachmentProgress", attachmentProgress, tabId);
  };

  insertImage = async (
    image: Omit<ImageAttributes, "bloburl"> & {
      dataurl: string;
    },
    tabId: string
  ) => {
    await this.sendCommand("insertImage", image, tabId);
  };

  handleBack = async () => {
    return this.sendCommand("handleBack");
  };

  keyboardShown = async (keyboardShown: boolean) => {
    return this.sendCommand("keyboardShown", keyboardShown);
  };

  getTableOfContents = async () => {
    const tabId = useTabStore.getState().currentTab;
    return this.sendCommand("getTableOfContents", tabId);
  };

  focusPassInput = async () => {
    const tabId = useTabStore.getState().currentTab;
    return this.sendCommand("focusPassInput", tabId);
  };

  blurPassInput = async () => {
    const tabId = useTabStore.getState().currentTab;
    return this.sendCommand("blurPassInput", tabId);
  };

  createInternalLink = async (
    attributes: LinkAttributes,
    resolverId: string
  ) => {
    if (!resolverId) return;
    return this.sendCommand("createInternalLink", attributes, resolverId);
  };

  dismissCreateInternalLinkRequest = async (resolverId: string) => {
    if (!resolverId) return;
    return this.sendCommand("dismissCreateInternalLinkRequest", resolverId);
  };

  scrollIntoViewById = async (id: string) => {
    const tabId = useTabStore.getState().currentTab;
    return this.sendCommand("scrollIntoViewById", id, tabId);
  };

  scrollToSearchResult = (index: number) => {
    const tabId = useTabStore.getState().currentTab;
    return this.sendCommand("scrollToSearchResult", index, tabId);
  };
}

export default Commands;
