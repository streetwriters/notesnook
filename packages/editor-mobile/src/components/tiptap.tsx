/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
import { Editor, TiptapOptions, Toolbar, useTiptap } from "@notesnook/editor";
import { useEffect } from "react";
import { useTabContext } from "../hooks/useTabStore";
import { EmotionEditorToolbarTheme } from "../theme-factory";
import { Settings } from "../utils";
export default function TiptapEditorWrapper(props: {
  options: Partial<TiptapOptions>;
  onEditorUpdate: (editor: Editor) => void;
  settings: Settings;
}) {
  const tab = useTabContext();
  const editor = useTiptap(props.options, [props.options]);
  globalThis.editors[tab.id] = editor;

  useEffect(() => {
    props.onEditorUpdate(editor);
  }, [editor, props]);

  return (
    <>
      {tab.session?.locked ? null : (
        <EmotionEditorToolbarTheme>
          <Toolbar
            className="theme-scope-editorToolbar"
            sx={{
              display: props.settings.noToolbar ? "none" : "flex",
              overflowY: "hidden",
              minHeight: "50px"
            }}
            editor={editor}
            location="bottom"
            tools={
              Array.isArray(props.settings.tools)
                ? [...props.settings.tools]
                : []
            }
            defaultFontFamily={props.settings.fontFamily}
            defaultFontSize={props.settings.fontSize}
          />
        </EmotionEditorToolbarTheme>
      )}
    </>
  );
}
