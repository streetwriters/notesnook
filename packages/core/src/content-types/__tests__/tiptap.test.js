/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import {
  IMG_CONTENT,
  IMG_CONTENT_WITHOUT_HASH
} from "../../../__tests__/utils/index.ts";
import { Tiptap } from "../tiptap.ts";
import { test, expect } from "vitest";

test("img src is empty after extract attachments", async () => {
  const tiptap = new Tiptap(IMG_CONTENT_WITHOUT_HASH);
  const result = await tiptap.postProcess(async () => {
    return "helloworld";
  });
  expect(result.hashes).toHaveLength(1);
  expect(result.data).not.toContain(`src="data:image/png;`);
  expect(result.data).not.toContain(`src=""`);
  expect(result.data).toContain(`data-hash="helloworld"`);
});

test("img src is present after insert attachments", async () => {
  const tiptap = new Tiptap(IMG_CONTENT);
  const result = await tiptap.postProcess(async () => {
    return { key: "hello", metadata: {} };
  });
  const tiptap2 = new Tiptap(result.data);
  const result2 = await tiptap2.insertMedia((hashes) => {
    const images = {};
    for (const hash of hashes) {
      images[hash] = "i am a data";
    }
    return images;
  });
  expect(result2).toContain(`src="i am a data"`);
});

test("remove attachments with particular hash", async () => {
  const tiptap = new Tiptap(IMG_CONTENT);
  const result = tiptap.removeAttachments(["d3eab72e94e3cd35"]);
  expect(result).not.toContain(`d3eab72e94e3cd35`);
});

const HTMLS = {
  tables: `<table style="border-collapse: collapse; width: 811px;" border="1"><tbody><tr><td style="width: 81.375px;">Goal</td><td style="width: 708.625px;">To introduce various features of the app to the user and to convert a user on trial or basic plan to upgrade.</td></tr><tr><td style="width: 81.375px;">Frequency</td><td style="width: 708.625px;">1/week or 2/week</td></tr><tr><td style="width: 81.375px;">Types</td><td style="width: 708.625px;">Feature intro, upgrade promo, one time emails</td></tr><tr><td style="width: 81.375px;"></td><td style="width: 708.625px;"></td></tr></tbody></table><h2>Emails</h2><h3>Feature intro</h3><p>Features:</p><ol style="list-style-type: decimal;"><li>Web clipper on mobile</li><li>Pin any note to notification</li><li>Take notes from notifications</li><li>App lock</li><li>Importer</li><li>Encrypted attachments</li><li>Session history &amp; automatic backups</li><li>Note publishing</li><li>Note exports</li><li>Collapsible headers</li></ol><h3>Promos</h3><ol style="list-style-type: decimal;"><li>Trial about to end</li><li>Trial ending (with option to request an extension)</li><li>Try free for 14 days</li></ol><h3>One time</h3><ol style="list-style-type: decimal;"><li>End-of-month progress report</li><li>What's coming/roadmap</li><li>What we are working on</li><li>Join the community</li></ol>`,
  tables2: `<h1>Note 8/6/22, 10:48 AM</h1>
  <p  data-spacing="double">hell</p><p  data-spacing="double">what</p><table><tbody><tr><th colspan="1" rowspan="1"><p  data-spacing="double">sdsdavav</p></th><th colspan="1" rowspan="1"><p  data-spacing="double">daskvjbdsva</p></th><th colspan="1" rowspan="1"><p  data-spacing="double">vsadjkvsadbvjk</p></th></tr><tr><td colspan="1" rowspan="1"><p  data-spacing="double">dsvsajkdb</p></td><td colspan="1" rowspan="1"><p  data-spacing="double">dskajvbsadj</p></td><td colspan="1" rowspan="1"><p  data-spacing="double">kjdasvbkj</p></td></tr><tr><td colspan="1" rowspan="1"><p  data-spacing="double">daskvbkdsa</p></td><td colspan="1" rowspan="1"><p  data-spacing="double">kdsajvbsajkd</p></td><td colspan="1" rowspan="1"><p  data-spacing="double">kjdsavbdsa</p></td></tr></tbody></table>`,
  codeblocks: `<p>Typescript is one of those languages that appear to be very simple. It's often described as "Javascript with types" and it fits that name very well. However, what many don't realize starting out with Typescript is that Typescript&nbsp;<em>is </em>a language and like all other languages it has it's own "secrets", it's own set of quirks.</p><p>When I started out with Typescript a few years back, I absolutely hated it. It was unnecessary, a box of clutter, making me write code that would never actually run. I hated defining interfaces, typing out all my functions, and thinking in terms I was not used to as a Javascript developer. Before Javascript, I had coded in C# and I had never really liked C# (I still don't) mostly for the huge amounts of boilerplate and magic involved. Typescript is heavily inspired by C# and seeing that contaminate the Javascript ecosystem irked me no end.</p><p>That was all just fluff. Typescript is a great language solving a considerable amount of problems for a huge amount of developers. And generics is the hidden weapon behind it all.</p><h2>Understanding Typescript generics</h2><p>For all points and purposes, generics shouldn't exist. They are one of the main factors behind abysmal code readability. But in the right hands, generics turn into a super-weapon.</p><p>The main problem generics try to solve is how to take in multiple types of parameters. This is a solved problem but requires duplicating your functions or using conditions to separate out different logic for different types. That is essentially what generics do as well but hidden from human eyes.</p><p>Think of a container that can take any type of item as long as it is not a circle. In Javascript, this becomes a problem you have to solve at runtime with checks &amp; conditions:</p><pre data-indent-type="space" data-indent-length="2" class="language-javascript"><code>var container = [];\nfunction putIntoContainer(item) {\n    if (item.type === "round") throw new Error("Rounded items not supported.")\n    container.push(item);\n}\n\nvar square = {type: "square"}\nvar circle = {type: "round"}\nputIntoContainer(square)\nputIntoContainer(circle) // ERROR! Rounded items not supported!\n</code></pre><p>There are many ways to solve this problem and some are even practical. The issue here isn't of repetition but of doing unnecessary work. Type safe languages would automatically give an error if there was a wrong type but Javascript knows nothing about the <code>item</code>.</p><p>In Typescript, this will be solved much more succinctly:</p><pre data-indent-type="space" data-indent-length="2" class="language-typescript"><code>// first define the types of items we'll handle\n// i.e. we don't want to handle any item other\n// than square or round.\n// This gives us nice auto completion and safety\n// against typos.\ntype ItemTypes = "square" | "round";\n\n// Define a generic item that can be of any type\n// defined in ItemTypes.\n// i.e. Item&lt;"triangle"&gt; will give an error.\ntype Item&lt;TItemType extends ItemTypes&gt; = {\n type?: TItemType;\n width: number;\n height: number;\n};\n\n// This is just syntax sugar to increase readability.\ntype Square = Item&lt;"square"&gt;;\ntype Circle = Item&lt;"round"&gt;;\n\n// Our container is just a simple wrapper around an array\n// that accepts items of only a specific type.\ntype Container&lt;TItemType extends ItemTypes&gt; = Array&lt;Item&lt;TItemType&gt;&gt;;\n\nvar squareContainer: Container&lt;"square"&gt; = [];\nvar roundContainer: Container&lt;"round"&gt; = [];\n\n// This wrapper is unnecessary, of course, because array.push\n// already does this. Only for demonstration purposes.\nfunction putIntoContainer&lt;\n TItemType extends ItemTypes,\n TItem extends Item&lt;TItemType&gt;\n&gt;(container: Container&lt;TItemType&gt;, item: TItem) {\n container.push(item);\n}\n\nvar square: Square = { width: 100, height: 200 };\nvar circle: Circle = { width: 200, height: 500 };\n\nputIntoContainer(squareContainer, square);\nputIntoContainer(roundContainer, circle);\nputIntoContainer(roundContainer, square); // Error: Argument of type 'Square' is not assignable to parameter of type 'Item&lt;"round"&gt;'.\nputIntoContainer(squareContainer, circle); // Error: Argument of type 'Circle' is not assignable to parameter of type 'Item&lt;"square"&gt;'.\n</code></pre><p>A lot more code, I know, and if you don't know how generics work that blob of code is utter nonsense. One of the main reasons I avoided Typescript for a long time. But look at the benefits:</p><ol style="list-style-type: decimal"><li><p >You have 100% compile-time type safety.</p></li><li><p >You can't put a round item in a square container (you will get compiler error).</p></li><li><p >You didn't write any extra runtime code.</p></li></ol><p>Expanding on point #3, after transpilation the above code will turn more-or-less into:</p><pre data-indent-type="space" data-indent-length="2" class="language-javascript"><code>var squareContainer = [];\nvar roundContainer = [];\n\nfunction putIntoContainer(container, item) {\n container.push(item);\n}\n\nvar square = { width: 100, height: 200 };\nvar circle = { width: 200, height: 500 };\n\nputIntoContainer(squareContainer, square);\nputIntoContainer(roundContainer, circle);\n</code></pre><p>This is the power of generics. More specifically, this is Typescript generics at a glance.</p><p>But this post was supposed to be about the "Secrets" of Typescript Generics, right? Well, let's get into that.</p><h2>1. Type filters using ternary operators</h2><h2>2. Deeply recursive types</h2><h2>3. Type functions</h2><h2>4. Type inference using interface properties</h2>`,
  tasklists: `<p>Hello</p><ul data-collapsed="false" class="checklist"><li class="checked checklist--item"><p>Task item 1</p></li><li class="checked checklist--item"><p>Task item 2</p></li><li class="checked checklist--item"><p>Task item 3</p></li><li class="checklist--item"><p>Task item 4</p><ul data-collapsed="false" class="checklist"><li class="checklist--item"><p>Sub task item 1</p></li><li class="checklist--item"><p>Sub task item 2</p></li></ul></li><li class="checklist--item"><p>Task Item 5</p></li></ul><p>Nene</p><ul><li><p>dasvsadv</p></li><li><p>adsva\`sd</p></li><li><p>vasd</p></li><li><p>vsadvdsa</p></li></ul>`,
  outlinelists: `<p >Testing outline list:</p><ul data-collapsed="false" data-type="outlineList"><li data-type="outlineListItem"><p >My outline list</p></li><li data-type="outlineListItem"><p >works</p></li><li data-type="outlineListItem"><p >but sometimes</p><ul data-collapsed="false" data-type="outlineList"><li data-type="outlineListItem"><p >It doesn't</p></li><li data-type="outlineListItem"><p >what do I do?</p></li><li data-type="outlineListItem"><p >I need to do something!</p></li></ul></li><li data-type="outlineListItem"><p >Makes no sense!</p></li><li data-type="outlineListItem"><p >Yes it doesn't!</p></li></ul>`,
  codeblock2: `<pre>hello<br></pre>`,
  singleSpacedParagraphs: `<p data-spacing="single">hello world</p><p data-spacing="single">hello world 2</p>`
};

for (const html in HTMLS) {
  test(`convert HTML to markdown with ${html}`, () => {
    const tiptap = new Tiptap(HTMLS[html]);
    expect(tiptap.toMD()).toMatchSnapshot(`html-to-md-${html}.md`);
  });
}

for (const html in HTMLS) {
  test(`convert HTML to text with ${html}`, () => {
    const tiptap = new Tiptap(HTMLS[html]);
    expect(tiptap.toTXT()).toMatchSnapshot(`html-to-txt-${html}.txt`);
  });
}
