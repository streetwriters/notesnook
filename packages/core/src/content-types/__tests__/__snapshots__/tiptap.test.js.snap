// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convert HTML to markdown with codeblock2 > html-to-md-codeblock2.md 1`] = `
"\`\`\`
hello
\`\`\`

"
`;

exports[`convert HTML to markdown with codeblocks > html-to-md-codeblocks.md 1`] = `
"Typescript is one of those languages that appear to be very simple. It's often described as "Javascript with types" and it fits that name very well. However, what many don't realize starting out with Typescript is that Typescript*is *a language and like all other languages it has it's own "secrets", it's own set of quirks.

When I started out with Typescript a few years back, I absolutely hated it. It was unnecessary, a box of clutter, making me write code that would never actually run. I hated defining interfaces, typing out all my functions, and thinking in terms I was not used to as a Javascript developer. Before Javascript, I had coded in C# and I had never really liked C# (I still don't) mostly for the huge amounts of boilerplate and magic involved. Typescript is heavily inspired by C# and seeing that contaminate the Javascript ecosystem irked me no end.

That was all just fluff. Typescript is a great language solving a considerable amount of problems for a huge amount of developers. And generics is the hidden weapon behind it all.

## Understanding Typescript generics

For all points and purposes, generics shouldn't exist. They are one of the main factors behind abysmal code readability. But in the right hands, generics turn into a super-weapon.

The main problem generics try to solve is how to take in multiple types of parameters. This is a solved problem but requires duplicating your functions or using conditions to separate out different logic for different types. That is essentially what generics do as well but hidden from human eyes.

Think of a container that can take any type of item as long as it is not a circle. In Javascript, this becomes a problem you have to solve at runtime with checks & conditions:

\`\`\`javascript
var container = [];
function putIntoContainer(item) {
    if (item.type === "round") throw new Error("Rounded items not supported.")
    container.push(item);
}

var square = {type: "square"}
var circle = {type: "round"}
putIntoContainer(square)
putIntoContainer(circle) // ERROR! Rounded items not supported!
\`\`\`

There are many ways to solve this problem and some are even practical. The issue here isn't of repetition but of doing unnecessary work. Type safe languages would automatically give an error if there was a wrong type but Javascript knows nothing about the \`item\`.

In Typescript, this will be solved much more succinctly:

\`\`\`typescript
// first define the types of items we'll handle
// i.e. we don't want to handle any item other
// than square or round.
// This gives us nice auto completion and safety
// against typos.
type ItemTypes = "square" | "round";

// Define a generic item that can be of any type
// defined in ItemTypes.
// i.e. Item<"triangle"> will give an error.
type Item<TItemType extends ItemTypes> = {
 type?: TItemType;
 width: number;
 height: number;
};

// This is just syntax sugar to increase readability.
type Square = Item<"square">;
type Circle = Item<"round">;

// Our container is just a simple wrapper around an array
// that accepts items of only a specific type.
type Container<TItemType extends ItemTypes> = Array<Item<TItemType>>;

var squareContainer: Container<"square"> = [];
var roundContainer: Container<"round"> = [];

// This wrapper is unnecessary, of course, because array.push
// already does this. Only for demonstration purposes.
function putIntoContainer<
 TItemType extends ItemTypes,
 TItem extends Item<TItemType>
>(container: Container<TItemType>, item: TItem) {
 container.push(item);
}

var square: Square = { width: 100, height: 200 };
var circle: Circle = { width: 200, height: 500 };

putIntoContainer(squareContainer, square);
putIntoContainer(roundContainer, circle);
putIntoContainer(roundContainer, square); // Error: Argument of type 'Square' is not assignable to parameter of type 'Item<"round">'.
putIntoContainer(squareContainer, circle); // Error: Argument of type 'Circle' is not assignable to parameter of type 'Item<"square">'.
\`\`\`

A lot more code, I know, and if you don't know how generics work that blob of code is utter nonsense. One of the main reasons I avoided Typescript for a long time. But look at the benefits:

1. You have 100% compile-time type safety.

2. You can't put a round item in a square container (you will get compiler error).

3. You didn't write any extra runtime code.

Expanding on point #3, after transpilation the above code will turn more-or-less into:

\`\`\`javascript
var squareContainer = [];
var roundContainer = [];

function putIntoContainer(container, item) {
 container.push(item);
}

var square = { width: 100, height: 200 };
var circle = { width: 200, height: 500 };

putIntoContainer(squareContainer, square);
putIntoContainer(roundContainer, circle);
\`\`\`

This is the power of generics. More specifically, this is Typescript generics at a glance.

But this post was supposed to be about the "Secrets" of Typescript Generics, right? Well, let's get into that.

## 1\\. Type filters using ternary operators

## 2\\. Deeply recursive types

## 3\\. Type functions

## 4\\. Type inference using interface properties

"
`;

exports[`convert HTML to markdown with leadingWhitespaces > html-to-md-leadingWhitespaces.md 1`] = `
"two spaces

four spaces  
two spaces  
four spaces

"
`;

exports[`convert HTML to markdown with outlinelists > html-to-md-outlinelists.md 1`] = `
"Testing outline list:

- My outline list

- works

- but sometimes

    - It doesn't

    - what do I do?

    - I need to do something!

- Makes no sense!

- Yes it doesn't!

"
`;

exports[`convert HTML to markdown with singleSpacedParagraphs > html-to-md-singleSpacedParagraphs.md 1`] = `
"hello world  
hello world 2

"
`;

exports[`convert HTML to markdown with tables > html-to-md-tables.md 1`] = `
"| Goal                                                                                                          | To introduce various features of the app to the user and to convert a user on trial or basic plan to upgrade. |
| ------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------- |
| Frequency                                                                                                     | 1/week or 2/week                                                                                              |
| Types                                                                                                         | Feature intro, upgrade promo, one time emails                                                                 |
|                                                                                                               |                                                                                                               |

## Emails

### Feature intro

Features:

1. Web clipper on mobile
2. Pin any note to notification
3. Take notes from notifications
4. App lock
5. Importer
6. Encrypted attachments
7. Session history & automatic backups
8. Note publishing
9. Note exports
10. Collapsible headers

### Promos

1. Trial about to end
2. Trial ending (with option to request an extension)
3. Try free for 14 days

### One time

1. End-of-month progress report
2. What's coming/roadmap
3. What we are working on
4. Join the community

"
`;

exports[`convert HTML to markdown with tables2 > html-to-md-tables2.md 1`] = `
"# Note 8/6/22, 10:48 AM

hell

what

| sdsdavav       | daskvjbdsva    | vsadjkvsadbvjk |
| -------------- | -------------- | -------------- |
| dsvsajkdb      | dskajvbsadj    | kjdasvbkj      |
| daskvbkdsa     | kdsajvbsajkd   | kjdsavbdsa     |

"
`;

exports[`convert HTML to markdown with tasklists > html-to-md-tasklists.md 1`] = `
"Hello

- [x] Task item 1

- [x] Task item 2

- [x] Task item 3

- [ ] Task item 4

    - [ ] Sub task item 1

    - [ ] Sub task item 2

- [ ] Task Item 5

Nene

- dasvsadv

- adsva\\\`sd

- vasd

- vsadvdsa

"
`;

exports[`convert HTML to text with codeblock2 > html-to-txt-codeblock2.txt 1`] = `
"hello
"
`;

exports[`convert HTML to text with codeblocks > html-to-txt-codeblocks.txt 1`] = `
"Typescript is one of those languages that appear to be very simple. It's often
described as "Javascript with types" and it fits that name very well. However,
what many don't realize starting out with Typescript is that TypescriptÂ is a
language and like all other languages it has it's own "secrets", it's own set of
quirks.

When I started out with Typescript a few years back, I absolutely hated it. It
was unnecessary, a box of clutter, making me write code that would never
actually run. I hated defining interfaces, typing out all my functions, and
thinking in terms I was not used to as a Javascript developer. Before
Javascript, I had coded in C# and I had never really liked C# (I still don't)
mostly for the huge amounts of boilerplate and magic involved. Typescript is
heavily inspired by C# and seeing that contaminate the Javascript ecosystem
irked me no end.

That was all just fluff. Typescript is a great language solving a considerable
amount of problems for a huge amount of developers. And generics is the hidden
weapon behind it all.


UNDERSTANDING TYPESCRIPT GENERICS

For all points and purposes, generics shouldn't exist. They are one of the main
factors behind abysmal code readability. But in the right hands, generics turn
into a super-weapon.

The main problem generics try to solve is how to take in multiple types of
parameters. This is a solved problem but requires duplicating your functions or
using conditions to separate out different logic for different types. That is
essentially what generics do as well but hidden from human eyes.

Think of a container that can take any type of item as long as it is not a
circle. In Javascript, this becomes a problem you have to solve at runtime with
checks & conditions:

var container = [];
function putIntoContainer(item) {
    if (item.type === "round") throw new Error("Rounded items not supported.")
    container.push(item);
}

var square = {type: "square"}
var circle = {type: "round"}
putIntoContainer(square)
putIntoContainer(circle) // ERROR! Rounded items not supported!


There are many ways to solve this problem and some are even practical. The issue
here isn't of repetition but of doing unnecessary work. Type safe languages
would automatically give an error if there was a wrong type but Javascript knows
nothing about the item.

In Typescript, this will be solved much more succinctly:

// first define the types of items we'll handle
// i.e. we don't want to handle any item other
// than square or round.
// This gives us nice auto completion and safety
// against typos.
type ItemTypes = "square" | "round";

// Define a generic item that can be of any type
// defined in ItemTypes.
// i.e. Item<"triangle"> will give an error.
type Item<TItemType extends ItemTypes> = {
 type?: TItemType;
 width: number;
 height: number;
};

// This is just syntax sugar to increase readability.
type Square = Item<"square">;
type Circle = Item<"round">;

// Our container is just a simple wrapper around an array
// that accepts items of only a specific type.
type Container<TItemType extends ItemTypes> = Array<Item<TItemType>>;

var squareContainer: Container<"square"> = [];
var roundContainer: Container<"round"> = [];

// This wrapper is unnecessary, of course, because array.push
// already does this. Only for demonstration purposes.
function putIntoContainer<
 TItemType extends ItemTypes,
 TItem extends Item<TItemType>
>(container: Container<TItemType>, item: TItem) {
 container.push(item);
}

var square: Square = { width: 100, height: 200 };
var circle: Circle = { width: 200, height: 500 };

putIntoContainer(squareContainer, square);
putIntoContainer(roundContainer, circle);
putIntoContainer(roundContainer, square); // Error: Argument of type 'Square' is not assignable to parameter of type 'Item<"round">'.
putIntoContainer(squareContainer, circle); // Error: Argument of type 'Circle' is not assignable to parameter of type 'Item<"square">'.


A lot more code, I know, and if you don't know how generics work that blob of
code is utter nonsense. One of the main reasons I avoided Typescript for a long
time. But look at the benefits:
 1. You have 100% compile-time type safety.
 2. You can't put a round item in a square container (you will get compiler
    error).
 3. You didn't write any extra runtime code.

Expanding on point #3, after transpilation the above code will turn more-or-less
into:

var squareContainer = [];
var roundContainer = [];

function putIntoContainer(container, item) {
 container.push(item);
}

var square = { width: 100, height: 200 };
var circle = { width: 200, height: 500 };

putIntoContainer(squareContainer, square);
putIntoContainer(roundContainer, circle);


This is the power of generics. More specifically, this is Typescript generics at
a glance.

But this post was supposed to be about the "Secrets" of Typescript Generics,
right? Well, let's get into that.


1. TYPE FILTERS USING TERNARY OPERATORS


2. DEEPLY RECURSIVE TYPES


3. TYPE FUNCTIONS


4. TYPE INFERENCE USING INTERFACE PROPERTIES"
`;

exports[`convert HTML to text with leadingWhitespaces > html-to-txt-leadingWhitespaces.txt 1`] = `
"Â Â two spaces

Â Â Â Â four spaces
Â Â two spaces
Â Â Â Â four spaces"
`;

exports[`convert HTML to text with outlinelists > html-to-txt-outlinelists.txt 1`] = `
"Testing outline list:
 * My outline list
 * works
 * but sometimes
   * It doesn't
   * what do I do?
   * I need to do something!
 * Makes no sense!
 * Yes it doesn't!"
`;

exports[`convert HTML to text with singleSpacedParagraphs > html-to-txt-singleSpacedParagraphs.txt 1`] = `
"hello world
hello world 2"
`;

exports[`convert HTML to text with tables > html-to-txt-tables.txt 1`] = `
"Goal        To introduce various features of the app to the user and to
            convert a user on trial or basic plan to upgrade.
Frequency   1/week or 2/week
Types       Feature intro, upgrade promo, one time emails
            


EMAILS


FEATURE INTRO

Features:
 1.  Web clipper on mobile
 2.  Pin any note to notification
 3.  Take notes from notifications
 4.  App lock
 5.  Importer
 6.  Encrypted attachments
 7.  Session history & automatic backups
 8.  Note publishing
 9.  Note exports
 10. Collapsible headers


PROMOS

 1. Trial about to end
 2. Trial ending (with option to request an extension)
 3. Try free for 14 days


ONE TIME

 1. End-of-month progress report
 2. What's coming/roadmap
 3. What we are working on
 4. Join the community"
`;

exports[`convert HTML to text with tables2 > html-to-txt-tables2.txt 1`] = `
"NOTE 8/6/22, 10:48 AM


hell

what

SDSDAVAV     DASKVJBDSVA    VSADJKVSADBVJK
dsvsajkdb    dskajvbsadj    kjdasvbkj
daskvbkdsa   kdsajvbsajkd   kjdsavbdsa"
`;

exports[`convert HTML to text with tasklists > html-to-txt-tasklists.txt 1`] = `
"Hello
 â Task item 1
 â Task item 2
 â Task item 3
 â Task item 4
   â Sub task item 1
   â Sub task item 2
 â Task Item 5

Nene
 * dasvsadv
 * adsva\`sd
 * vasd
 * vsadvdsa"
`;
